### PR TITLE
CI: Add a Cygwin run to GHA CI.

### PR DESCRIPTION
**Pull request by DWesl**
## PR Summary

Add a Cygwin run to GHA, as suggested by https://github.com/matplotlib/matplotlib/issues/22997#issuecomment-1120064419

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by oscargus** [comment]
I think that libjpeg-devel is also required (as you probably have figured out as well).

**Comment by DWesl** [comment]
I can try that and see if it works (I shouldn't run into python-pillow/pillow#6216)

**Comment by DWesl** [comment]
The build is now failing in the expected place, so let's see if the suggestion in https://github.com/matplotlib/matplotlib/issues/22997#issuecomment-1120195190 allows the build to finish.

**Comment by DWesl** [comment]
The build is succeeding now, but the tests fail with `ImageComparisonError`s due to using the wrong version of FreeType.  Let's see if that compiles now.

**Comment by DWesl** [comment]
The freetype compile still fails during configure; the closest Cygwin build script I can find is [here](https://cygwin.com/git-cygwin-packages/?p=git/cygwin-packages/freetype2.git;a=blob;f=freetype2.cygport;h=1e114909378ea1f823703d5d5c8e3f86a05578c5;hb=37487e0e76061cb37b260f3c6291e3b2e3578be4).  The first patch doesn't look like the kind of thing to fix a configure crash (probably the same as [this one](https://src.fedoraproject.org/rpms/freetype/raw/main/f/freetype-2.2.1-enable-valid.patch)), and the name of the second patch doesn't look relevant.  

The only other thing I can think of is running `autoconf`/`autoreconf`.

**Comment by DWesl** [comment]
The Cygwin CI is still failing in FreeType configure:
https://github.com/matplotlib/matplotlib/runs/6436739391?check_suite_focus=true#step:11:756

I can get the compilation done using system FreeType and QHull, but building the FreeType version needed for the tests to pass  fails, either during the `libtoolize` step of `autoreconf` or during configure without that step.

Is there a way to skip all the tests that require an exact image match?  It wouldn't provide a thorough test of Matplotlib on Cygwin, but most of the differences from Windows and Linux are in compiling and importing the extensions, so that might be enough.

**Comment by DWesl** [comment]
[The bundled FreeType build fails even if I push through after the `libtoolize` failure](https://github.com/matplotlib/matplotlib/runs/6793329589?check_suite_focus=true#step:11:12746).  I can try again without the bundled FreeType, and see if I can eliminate the `fork()` failures, even if the `ImageComparisonFailure`s  stick around.

**Comment by DWesl** [comment]
I don't know why upload-artifact fails; the previous step should give everyone read and write permissions to everything in that directory and also execute permissions to every sub-directory.

**Comment by DWesl** [comment]
I haven't seen that error message on a `fork()` failure before:
```
      1 [main] python3.8 (5008) C:\cygwin\bin\python3.8.exe: *** fatal error in forked process - MEM_COMMIT failed, Win32 error 1455
  13465 [main] python3.8 (5008) cygwin_exception::open_stackdumpfile: Dumping stack trace to python3.8.exe.stackdump
2003466 [main] python 2011 dofork: child -1 - forked process 5008 died unexpectedly, retry 0, exit code 0x100, errno 11
```

The upload-artifact step is also still failing, and I have no idea why.

**Comment by DWesl** [comment]
Do Windows `ReadAndExecute` permissions not imply `Read` or `ReadPermissions`?  I wouldn't think `upload-artifact` would need `Write`, `ExecuteFile`, or delete permissions.

**Comment by DWesl** [comment]
And now the FreeType build is failing with no apparent reason given.

**Comment by QuLogic** [comment]
I do not understand all the extra work that needs to be done here. On Cygwin x86_64, I installed Python 3.9, `gcc-c++`, `make`, and I think `python39-matplotlib`, and was able to build FreeType just fine (with #23066). I did have trouble with downloading qhull, which I have not yet investigated.

**Comment by DWesl** [comment]
I was also able to build FreeType locally, but it still fails in CI, for reasons the error message does not seem to explain.  Trying to narrow down the cause of that failure is the bulk of the commits.

**Comment by DWesl** [comment]
> `fatal error in forked process - MEM_COMMIT failed, Win32 error 1455`

Well that's a new kind of `fork()` failure.

**Comment by DWesl** [comment]
The most recent commit should have stopped trying to install wxPython, so I'm not sure what to do about the six failing `MPLBACKEND: wxagg` tests

**Comment by tacaswell** [comment]
From this cryptic (truncated!) message:

```
stderr = 'Traceback (most recent call last):\n  File "<string>", line 1, in <module>\n  File "/cygdrive/d/a/matplotlib/matplotl....8/site-packages/wx/core.py", line 12, in <module>\n    from ._core import *\nImportError: No such file or directory\n'
```

I think that wx is still installed enough to pass our "can we test this" filter (it looks like there is a `wx/core.py` is site-packages?!) but not enough to actually run (the `_core` (I assume) c-extensions are mssing).

https://github.com/matplotlib/matplotlib/blob/0aac9f17b9373f66c709f436149568bda69595b3/lib/matplotlib/tests/test_backends_interactive.py#L24-L68

This logic can not actually import the GUI toolkits because importing more than one can lead to all manner of trouble so we check with `find_spec` and then run the actual tests is a sub-process (sorry if I am over explaining obvious things).

File: .github/workflows/tests.yml
Comment (tacaswell) [inline .github/workflows/tests.yml:416]: try to purge it?

```suggestion
              echo 'PySide2 is not available'
          python -m pip uninstall wxpynton
```

**Comment by tacaswell** [comment]
It looks like it is coming from the system packages?

```
  15 install python3-wx-common                      4.0.7.post2-4               
...
 401 install python38-wx                            4.0.7.post2-4           
```

https://github.com/matplotlib/matplotlib/actions/runs/3606263992/jobs/6077436300#step:4:104
https://github.com/matplotlib/matplotlib/actions/runs/3606263992/jobs/6077436300#step:4:490

**Comment by tacaswell** [comment]
```
FAILED lib/matplotlib/tests/test_matplotlib.py::test_tmpconfigdir_warning - AssertionError: assert 'set the MPLCONFIGDIR' in ''
634
 +  where '' = CompletedProcess(args=['/usr/bin/python', '-c', 'import matplotlib'], returncode=0, stderr='').stderr
635
FAILED lib/matplotlib/tests/test_pyplot.py::test_pylab_integration - BlockingIOError: [Errno 11] Resource temporarily unavailable
636
FAILED lib/matplotlib/tests/test_simplification.py::test_throw_rendering_complexity_exceeded
```

The tempdir one is already skiped on nt, I assume cygwin has the same limitations and should be skipped? 


The rendering complexity one may be an actual bug that is testing that c code bails on too-big things.  https://github.com/matplotlib/matplotlib/issues/21789 seems related and the most recent PR to touch this code was https://github.com/matplotlib/matplotlib/pull/19343.

I have no good theory on the pylab one.

**Comment by DWesl** [comment]
> ```
> FAILED lib/matplotlib/tests/test_matplotlib.py::test_tmpconfigdir_warning - AssertionError: assert 'set the MPLCONFIGDIR' in ''
> 634
>  +  where '' = CompletedProcess(args=['/usr/bin/python', '-c', 'import matplotlib'], returncode=0, stderr='').stderr
> 635
> FAILED lib/matplotlib/tests/test_pyplot.py::test_pylab_integration - BlockingIOError: [Errno 11] Resource temporarily unavailable
> 636
> FAILED lib/matplotlib/tests/test_simplification.py::test_throw_rendering_complexity_exceeded
> ```
> 
> The tempdir one is already skipped on nt, I assume cygwin has the same limitations and should be skipped?

Cygwin implements its own permissions: to check I just ran the code from the test locally and got the expected warning.
Extrapolating from [this FAQ entry](https://cygwin.com/faq.html#faq.api.executables), I think whether the test could work would depend on whether the underlying filesystem is FAT/FAT32 (fails) or NTFS/NFS (can pass).  There's a possibility there's default ACLs on one of the parent directories ([implied by a different FAQ entry](https://cygwin.com/faq.html#faq.using.same-with-permissions)).

The other possibility is that the test for root isn't working since the user ID for "Administrator" even running as administrator is not 0 by default (see [this user guide description](https://cygwin.com/cygwin-ug-net/ntsec.html#ntsec-mapping)).  I could try changing that using [mkpasswd](https://cygwin.com/cygwin-ug-net/mkpasswd.html), or try to look up how to tell if the current session has the relevant "Access files regardless of permissions" token.
> 
> The rendering complexity one may be an actual bug that is testing that c code bails on too-big things. #21789 seems related and the most recent PR to touch this code was #19343.
> 
> I have no good theory on the pylab one.

That's a `fork()` failure, [reasonably well-known on Cygwin](https://cygwin.com/cygwin-ug-net/mkpasswd.html).  The three main recommendations are:
- stop antivirus software (probably not relevant on CI, also not something we can control on CI)
- rebase all DLLs so they have distinct memory regions to load to so they don't collide while Cygwin tries to recreate the parent memory layout in the child (done [here](https://github.com/matplotlib/matplotlib/pull/22999/files#diff-1db27d93186e46d3b441ece35801b244db8ee144ff1405ca27a163bfe878957fR442-R449) in this PR; I can try adding `/bin` to be absolutely sure all files have an address and that address is found
- restart the computer and hope it goes away (WIndows makes fork even harder sometimes; probably not relevant on CI)

**Comment by DWesl** [comment]
It looks like all the rest are fork failures.  There was only one before I tried to fix it, so let's see if undoing that fix gets things passing.

In case that doesn't work, is there some way to tell pytest that all instances of `BlockingIOError: [Errno 11] Resource temporarily unavailable` should be treated as a non-strict xfail of the relevant test on Cygwin?

**Comment by tacaswell** [comment]
That logic should probably go around the subprocess calls in the helper used in the sphinx tests + `subprocess_run_helper`?

Might also be worth putting wrapped versions of the `subproccess` functions we use an `matplotlib.testing` and then making sure we never directly use the stdlib functions?

**Comment by DWesl** [comment]
It looks like the failures are mostly docs-related, then a compilation failure on Windows.

**Comment by DWesl** [comment]
Should I have moved the stdint include after the Python.h include rather than moving Python.h before stdint.h?

File: lib/matplotlib/testing/__init__.py
Comment (tacaswell) [inline lib/matplotlib/testing/__init__.py:12]: burying this import in a function should fix the docs (pytest is not a build dependency, but we want to document things from this sub-package in the docs.

File: .github/workflows/tests.yml
Comment (tacaswell) [inline .github/workflows/tests.yml:500]:
```suggestion

      - name: Upload code coverage
        uses: codecov/codecov-action@v3


```

**Review by tacaswell — APPROVED** [review]
Mild preference that the codecov uploader goes in, but OK merging with out it.

**Comment by DWesl** [comment]
Should I squash the first ninety or so commits?  Most of them are changes to the debugging level, and one or two got merged in other PRs; it's mostly the last six or so that do other things.

**Comment by tacaswell** [comment]
On one hand, it is currently at a very nice number of commits (100 🤯), but on the other I do not think we need the full history and pushing this to one commit for the c change, one commit for the change to the tests and one commit for the GHA config is probably the right move...

We can also squash-merge.

**Comment by tacaswell** [comment]
I am inclined to squash-merge this as-is.

**Comment by DWesl** [comment]
Seems you aren't attached to the 100-commit bit, so here's a three-commit version.

**Comment by DWesl** [comment]
Looking at the skipped tests, several of the dependencies are available on Cygwin; I could try adding them and seeing what happens.  The TeX tests are likely to run into fork failures, as are the ImageMagick animation tests and those requiring GhostScript, but the Tk tests should work fine.

File: .github/workflows/tests.yml
Comment (QuLogic) [inline .github/workflows/tests.yml:478]:
```suggestion
          mkdir -p ~/.config/matplotlib/
          sed -e 's~#animation.ffmpeg_path:.\+~animation.ffmpeg_path: /usr/bin/ffmpeg.exe~' "${oldmplrc}" >~/.config/matplotlib/matplotlibrc
```

File: .github/workflows/tests.yml
Comment (QuLogic) [inline .github/workflows/tests.yml:483]: Why is this needed?

File: .github/workflows/tests.yml
Comment (QuLogic) [inline .github/workflows/tests.yml:499]: Is this because of the `root` stuff?

File: .github/workflows/tests.yml
Comment (QuLogic) [inline .github/workflows/tests.yml:506]: Debugging that can be deleted?

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:306]: A version; Cygwin has 3.5-3.9, though fewer packages are available on 3.5

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:350]: I'd prefer to xfail if it's a failure only when run as Administrator; the test in question is probably `lib/matplotlib/tests/test_matplotlib.py::test_tmpconfigdir_warning`.

If you know how to ask GitHub Actions to run the tests as a normal user, that would also work.  [This method](https://sourceware.org/legacy-ml/cygwin/2012-04/msg00628.html) should also work if that doesn't.

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:366]: I thought that might be why the FreeType builds were failing, and dash is said to be faster than bash.

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:358]: I'm not sure what this is referring to.

I have not gotten actions/checkout to work with Cygwin git; this is the NumPy solution to Windows git not telling Cygwin git about newly-cloned repositories.

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:407]: Mostly, yes

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:461]: It's a shell that understands UNIX syntax (such as the locations of the rebase binary and the file list) but does not link against `cygwin1.dll`, which would prevent it from being able to relocate that file.  I'm not confident enough in Windows `command.com` scripts nor Windows PowerShell to try rewriting it in those.

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:462]: If I were more confident in my Windows scripting abilities or in the lack of need to rebase `cygwin1.dll`, yes.

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:480]: Different attempt to fix the tempfile test; this can go.

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:496]: Maybe?  actions/upload-artifact fails, and printing the permissions with a Windows utility doesn't show many permissions.  (This was an attempt to make actions/upload-artifact succeed, but didn't work.  That was an attempt to debug the failures with system FreeType, now unneeded).

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:503]: Probably.

File: .github/workflows/tests.yml
Comment (QuLogic) [reply .github/workflows/tests.yml:365]: I was referring to the above user ID change; if this is unrelated to that, then my comment can be ignored.

File: .github/workflows/tests.yml
Comment (QuLogic) [reply .github/workflows/tests.yml:458]: That makes sense, assuming it's done in place or something. Please leave a comment here as to why it's `ash` instead of `bash`.

File: .github/workflows/tests.yml
Comment (QuLogic) [reply .github/workflows/tests.yml:373]: It would be odd, as `bash` is used to configure it on most Linux. If this does fix things though, it should have a comment.

File: .github/workflows/tests.yml
Comment (tacaswell) [inline .github/workflows/tests.yml:313]:
```suggestion
        python-minor-version: [8, 9]
```

Is this tempting fate too much?

File: .github/workflows/tests.yml
Comment (QuLogic) [reply .github/workflows/tests.yml:357]: Unfortunately, I don't have a good idea how either. Please leave a comment on this one as well.

**Comment by QuLogic** [comment]
Note, we do prefer rebases over merges, if you can.

**Comment by efiring** [comment]
It's not clear to me that this is worth the added CI load and complexity.

**Comment by QuLogic** [comment]
We discussed this a bit on call; there was some worry about how fragile Cygwin might be given none(?) of the core developers use it. I suggest making it less required by a) moving it to a separate file `cygwin.yml` (AFAIK, there's no real dependence on anything else in `tests.yml`); b) triggering on a schedule and label instead of pull request as is done in `cibuildwheel.yml`. I think we are unlikely to break Cygwin builds often unless editing the C++ files, so a bonus would be to get the build to automatically trigger if files in `src/` were changed, but that's just a nice-to-have.

**Comment by DWesl** [comment]
> Note, we do prefer rebases over merges, if you can.

Rebased and consolidated commits again.

> We discussed this a bit on call; there was some worry about how fragile Cygwin might be given none(?) of the core developers use it. I suggest making it less required by 
> a) moving it to a separate file `cygwin.yml` (AFAIK, there's no real dependence on anything else in `tests.yml`); 

Done

> b) triggering on a schedule and label instead of pull request as is done in `cibuildwheel.yml`. 

I set up conditions to mirror those in cibuildwheel.yml, though I kept the scheduled runs and "run when requested" bits.

> I think we are unlikely to break Cygwin builds often unless editing the C++ files, so a bonus would be to get the build to automatically trigger if files in `src/` were changed, but that's just a nice-to-have.

I think I have this in the conditions, though it looks like that one needs both conditions met rather than either.

File: .github/workflows/cygwin.yml
Comment (QuLogic) [inline .github/workflows/cygwin.yml:11]:
```suggestion
      - v[0-9]+.[0-9]+.[0-9x]+
```

File: .github/workflows/cygwin.yml
Comment (QuLogic) [inline .github/workflows/cygwin.yml:71]:
```suggestion
    strategy:
```

File: .github/workflows/cygwin.yml
Comment (QuLogic) [inline .github/workflows/cygwin.yml:160]:
```suggestion
            C:\cygwin\home\runneradmin\.cache\matplotlib
            !C:\cygwin\home\runneradmin\.cache\matplotlib\tex.cache
            !C:\cygwin\home\runneradmin\.cache\matplotlib\test_cache
```

File: .github/workflows/cygwin.yml
Comment (QuLogic) [inline .github/workflows/cygwin.yml:166]:
```suggestion
      - name: Ensure correct Python version
```

File: .github/workflows/cygwin.yml
Comment (QuLogic) [inline .github/workflows/cygwin.yml:195]:
```suggestion
      - name: Install Matplotlib
```

**Review by QuLogic — APPROVED** [review]
Minor typos.

**Comment by QuLogic** [comment]
Seems to be a typo:
https://github.com/matplotlib/matplotlib/actions/runs/3864270329

File: lib/matplotlib/testing/__init__.py
Comment (QuLogic) [inline lib/matplotlib/testing/__init__.py:56]: `float` is the type, not the default?

**Comment by tacaswell** [comment]
Lets not backport this.

**Comment by tacaswell** [comment]
This has two approval and is green so I'm going to merge!

**Comment by tacaswell** [comment]
Thank you for sticking with us @DWesl

**Pull request merged**
